CARDS-2585: More consistent treatment of out-of-range numerical answers and non-regexp-matching text answers

veronikaslc opened the pull request:
Specs: [CARDS-2585](https://phenotips.atlassian.net/issues/CARDS-2585)

[CARDS-2585]: https://phenotips.atlassian.net/browse/CARDS-2585?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
CARDS-1761: Adapt React Date / Time pickers
- Remaned DateQuestionUtils to TimeQuestionUtils;
- Adapted DateTimePicker to the Question components;
- Refactored DateQuestionMonth and DateQuestionFull components into one DateQuestion component;
- Adapted DatePicker in the DowntimeWarningConfiguration and PatientIdentification components;
- Adapted DateTimePicker in the DateFilter component;

veronikaslc committed:
CARDS-1979: Add ability to CLEAR the value of a date/time input in the date/time picker

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Moved DateTimeUtilities to commons package

veronikaslc committed:
[misc] Use checkPropTypes wrapper

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Fixed naming after rebase

veronikaslc committed:
CARDS-2585: More consistent treatment of out-of-range numerical answers and non-regexp-matchin text answers

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Question.jsx:
```diff
@@ -129,8 +129,9 @@ function Question (props) {
             ( labels ?
               <List>
                 { labels.map( (item, idx) => {
+                  let value = defaultDisplayFormatter ? defaultDisplayFormatter(item, idx) : item;
```
This would also work:

```suggestion
                  let value = defaultDisplayFormatter?.(item, idx) ?? item;
```

marta- replied:
However if the formatter decides to display `null` we fall back to `item`, which changes the original behavior. Please disregard.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Question.jsx:
```diff
@@ -129,8 +129,9 @@ function Question (props) {
+                  let value = defaultDisplayFormatter ? defaultDisplayFormatter(item, idx) : item;
                   return(
-                    <ListItem key={existingAnswer[0] + idx}> {defaultDisplayFormatter ? defaultDisplayFormatter(item, idx) : item}</ListItem>
+                    value ? <ListItem key={existingAnswer[0] + idx}> {value}</ListItem> : null
```
If `value` is `0`, nothing gets displayed. This is not pretty, but catches this one case:
```suggestion
                    value || value === 0 ? <ListItem key={existingAnswer[0] + idx}> {value}</ListItem> : null
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -410,7 +410,7 @@ function MultipleChoice(props) {
             helperText={
               inputError
-              ? <FormattedText variant="caption">{ validationErrorText }</FormattedText>
+              ? <><FormattedText variant="caption">{ validationErrorText }</FormattedText></>
```
~Why do we need to wrap in a react fragment?~

How does wrapping in a react fragment avoid div being descendent of p?

veronikaslc committed:
CARDS-2585: More consistent treatment of out-of-range numerical answers and non-regexp-matchin text answers

Better code logic

Co-authored-by: Marta Girdea <651980+marta-@users.noreply.github.com>

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/NumberQuestion.jsx:
```diff
@@ -302,9 +302,16 @@ function NumberQuestion(props) {
-      <FormattedText>
-        { `${initialValue?.[0]} &mdash; ${label}` }
-      </FormattedText>
+      <div>
```
Would `<>` instead of `<div>` work, to simplify the DOM?

veronikaslc replied:
It wouldn't

veronikaslc replied:
![image](https://github.com/user-attachments/assets/d8fbdd05-ffdf-49d7-b225-f289c6f0bf3b)

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Deleted DropdownsDatePicker

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/NumberQuestion.jsx:
```diff
@@ -302,9 +302,16 @@ function NumberQuestion(props) {
+        </FormattedText>
+        { pageActive && (minMaxError || rangeError) &&
+          <Typography component="div" color="error" variant="caption">
+            Invalid input
```
"Invalid input" is too vague, we have the ability to provide better feedback to the user as we know exactly what went wrong.  For `minMaxError`, we should say, depending on which limit is violated: "The values must be lower than ${maxValue} | greater than ${minValue} | between ${minValue} and ${maxValue". For `rangeError`, we can just use `rangeErrorMessage`.

veronikaslc replied:
I'll refactor a component abit to get the message at the validation stage.

marta- replied:
> depending on which limit is violated

Just to make sure I was clear: I'm not suggesting to use the same message as the answer instructions at line 281. That message, used in edit mode and displayed all the time, gives guidance about what values to enter.

This message, used in view mode and displayed only if the answer is invalid, should only mention the limits that are not met.

For example, for $minValue = 50 and $maxValue = 100, an answer of 500 should trigger the message "The values must be lower than 100". No need to mention minValue in view mode, since that is already satisfied.

veronikaslc replied:
Got it

veronikaslc replied:
Just a side note: we might have a problem here at line 160:
`const initialValue = Array.from(existingAnswer?.[1]?.value || []);`
If `existingAnswer?.[1]?.value` is a number the expression always return `[]`

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/AnswerInstructions.jsx:
```diff
@@ -59,7 +59,6 @@ function AnswerInstructions (props) {
 
   return (instructionsExist && (
     <Typography
-      component="p"
```
By removing this line, the default component (span) is used. However, span is an inline element and the styling rule which applies the negative top margin to compensate for the card content padding has no effect.

![image](https://github.com/user-attachments/assets/9926d26d-8e93-4ce0-8c5f-211e83c90e79)

So now we have this big gap:
![image](https://github.com/user-attachments/assets/c069a78b-ed6a-47b5-953f-2ca60e20dd8c)

Instead, it should look like this:
![image](https://github.com/user-attachments/assets/8869b6e3-1a40-405a-8237-044686e181a0)

I'm also not sure what problem gets solved by removing this `p`.  It's a `p` inside a `div` (which is permitted), not viceversa.

veronikaslc replied:
Here is what happens if I put -1 in place of -24
![image](https://github.com/user-attachments/assets/90a24abf-617b-4280-9687-3c4f26e1d355)

veronikaslc replied:
Reverted that change here [7ba26d7](https://github.com/data-team-uhn/cards/pull/1980/commits/7ba26d7b8d28a3c044d7739ba191f66a0a182ec8)

veronikaslc committed:
CARDS-2585: More consistent treatment of out-of-range numerical answers and non-regexp-matchin text answers
Reverted the p component for Typography removal

veronikaslc committed:
[misc] Fixed nested paragraph rendering and range padding

veronikaslc committed:
CARDS-2585: More consistent treatment of out-of-range numerical answers and non-regexp-matchin text answers
Custom message per validation error in view mode

veronikaslc committed:
[misc] Fixed clearing number range input

veronikaslc committed:
[misc] Fixed displaying saved 0 number input value in edit mode

veronikaslc committed:
CARDS-2585: More consistent treatment of out-of-range numerical answers and non-regexp-matching text answers
Better number error messages

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/NumberQuestion.jsx:
```diff
@@ -195,17 +195,17 @@ function NumberQuestion(props) {
       // Test that it is an integer
       if (!/^[-+]?\d*$/.test(text)) {
-        return true;
+        return "The values must be an integer";
```
Should we switch plurality here based on whether it's a range or not? Also, since we're tweaking these messages anyway, we might want to use the term "whole number" instead of the technical "integer".

```suggestion
        return isRange ? "The values must be whole numbers" : "The value must be a whole number";
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/NumberQuestion.jsx:
```diff
@@ -214,25 +214,31 @@ function NumberQuestion(props) {
       // Reject whitespace and non-numbers
       if (/^\s*$/.test(text) || isNaN(value)) {
-        return true;
+        return "The values must be a number";
```
```suggestion
        return isRange ? "The values must be numeric" : "The value must be numeric";
```

veronikaslc commented:
rebased on the latest dev, ready for test

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/NumberQuestion.jsx:
```diff
@@ -195,17 +195,17 @@ function NumberQuestion(props) {
       // Test that it is an integer
       if (!/^[-+]?\d*$/.test(text)) {
-        return true;
+        return `The value${isRange ? 's' : ''} must be whole numbers`;
```
`numbers` ->  `number${isRange ? 's' : ''}`

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/NumberQuestion.jsx:
```diff
@@ -214,25 +214,31 @@ function NumberQuestion(props) {
-    if ((typeof minValue !== 'undefined' && value < minValue) ||
-      (typeof maxValue !== 'undefined' && value > maxValue)) {
-      return true;
+    if ((typeof minValue !== 'undefined' && (isRange ? lowerLimit : value) < minValue) &&
```
The architecture of this method is awkward. We pass it `value` (via the `text` argument), but if `isRange` we completely ignore `value` on this block and check the upper/lower limits instead. If two values need to be checked at a time, we should pass them both as arguments.

veronikaslc replied:
The logic is unusual at a first glance, but in the case of  `isRange` the `value` is either `lower` or `upper` `limit` passed in lines 240-241. This way the changes to the code are minimal and at the same time no further customisation is needed to check one or two values. Besides, we always check them one by one for being numeric.

veronikaslc committed:
CARDS-2585: More consistent treatment of out-of-range numerical answers and non-regexp-matching text answers
Better error wording

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Display date error messages in edit and view mode

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers

Add comments

Co-authored-by: acrowthe <acrowthe3388@gmail.com>

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers

fix date format case

Co-authored-by: acrowthe <acrowthe3388@gmail.com>

acrowthe commented on modules/data-entry/src/main/frontend/src/questionnaire/NumberQuestion.jsx:
```diff
@@ -213,25 +213,31 @@ function NumberQuestion(props) {
-      return true;
+    if ((typeof minValue !== 'undefined' && (isRange ? lowerLimit : value) < minValue) &&
+        (typeof maxValue !== 'undefined' && (isRange ? upperLimit : value) > maxValue)) {
+      return `The value${isRange ? 's' : ''} must be between ${minValue} and ${maxValue}`;
```
This condition should only be possible if this is a range (since it needs to be both higher and lower than the limits) so can be simplified:
```suggestion
          if (isRange && typeof minValue !== 'undefined' && lowerLimit < minValue &&
            typeof maxValue !== 'undefined' && upperLimit > maxValue) {
            return `The values must be between ${minValue} and ${maxValue}`;
```

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Addressed comments

veronikaslc committed:
CARDS-2585: More consistent treatment of out-of-range numerical answers and non-regexp-matching text answers
Simplify the error check logic in one of steps

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
More consistent error validation on manual and picker date manipulation

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Added milliseconds to the default sling format as well as truncating

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Added answer instructions to the top

acrowthe commented on modules/data-entry/src/main/frontend/src/questionnaire/NumberQuestion.jsx:
```diff
@@ -301,22 +309,50 @@ function NumberQuestion(props) {
+        <FormattedText color={!disableMinMaxValueEnforcement && pageActive && minMaxError ? "error" : ""}>
+          { label }
+        </FormattedText>
+        { (disableMinMaxValueEnforcement && typeof messageForValuesOutsideMinMax != "undefined") ?
```
This leads to `messageForValuesOutsideMinMax` being displayed at all times if this message is defined and `disableMinMaxValueEnforcement` is defined as true. This message should only be displayed if the value is outside of the defined min and max.

Testing: `Number Min / Max Value Test` form, enter a valid number in the second question, save and view

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Addressed codereview comments

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Removed leftover code from using moment; Fixed saving empty date for DowntimeWarningConfiguration

veronikaslc committed:
CARDS-2585: More consistent treatment of out-of-range numerical answers and non-regexp-matching text answers
Addressed codereview comments

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Removed duplicated methods from DateTimeUtilities
Moved non-utilities methods from DateTimeUtilities to DateQuestion
Optimised answerType calculation in ComputedQuestion

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Use DateTimeUtilities formatDateAnswer() method to display date string

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Do not strip the timezone for the date processing

veronikaslc committed:
CARDS-1761: Adapt React Date / Time pickers
Correct meridiem/hour detection

sdumitriu committed:
Merge pull request #1929 from data-team-uhn/CARDS-1761-new

CARDS-1761: Adapt React Date / Time pickers
